### PR TITLE
F18 makefile for in-tree and out-of-tree builds.

### PR DESCRIPTION
**Pull request by sscalpone**
Use the add_llvm_library feature LINK_LIBS to inherit the library attribute, e.g. shared.

The PROJECT_SOURCE_DIR changes for in- and out-of-tree builds.  Now, find the module files relative to the tools/f18 directory.

### CODE REVIEW

File: tools/f18/CMakeLists.txt [hunk 1]
```diff
@@ -52,14 +52,14 @@ set(MODULES
 # Create module files directly from the top-level module source directory
 foreach(filename ${MODULES})
   add_custom_command(OUTPUT ${CMAKE_CURRENT_BINARY_DIR}/include/${filename}.mod
-    COMMAND f18 -fparse-only -fdebug-semantics ${PROJECT_SOURCE_DIR}/../flang/module/${filename}.f90
+    COMMAND f18 -fparse-only -fdebug-semantics ${CMAKE_CURRENT_SOURCE_DIR}/../../module/${filename}.f90
```
Comment (schweitzpgi) [inline tools/f18/CMakeLists.txt]: I don't understand why we're using hard coded relative paths, but this is an improvement.

We should have a FLANG_SOURCE_DIR variable being set someplace to make this more future-proof.

**Comment by schweitzpgi** [comment]
It's looking good. Thanks for the help.

File: lib/fir/CMakeLists.txt [hunk 2]
```diff
@@ -24,11 +24,7 @@ add_llvm_library(FIR
   StdConverter.cpp
   Tilikum.cpp
   Type.cpp
-)
-
-add_dependencies(FIR FIROpsIncGen)
-
-target_link_libraries(FIR
+  LINK_LIBS
```
[the remark this answers is not in the document]
Comment (sscalpone) [reply lib/fir/CMakeLists.txt]: CMake complains if the library being built has a different libtype from the libraries that it is using; using LINK_LIBS inherits the libtype in target_link_libraries.

[the remark this answers is not in the document]
Comment (schweitzpgi) [reply lib/fir/CMakeLists.txt]: Thanks. That's interesting. Why is CMake fine with the entire LLVM source tree (with scant exceptions) using `target_link_libraries` and only complains in f18?

**Comment by schweitzpgi** [comment]
This got really bent with the last rebase.

Can you send me a pointer to your scripts so I can test it?

**Pull request closed without merging**